sw/device/lib/crypto/impl/kdf.c

```diff
@@ -28,8 +29,109 @@ otcrypto_status_t otcrypto_kdf_kmac(
     otcrypto_kmac_mode_t kmac_mode, const otcrypto_const_byte_buf_t kdf_label,
     const otcrypto_const_byte_buf_t kdf_context, size_t required_byte_len,
     otcrypto_blinded_key_t *keying_material) {
-  // TODO: Implement KMAC-KDF.
-  return OTCRYPTO_NOT_IMPLEMENTED;
+  // Check NULL pointers.
+  if (key_derivation_key.keyblob == NULL || keying_material == NULL) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Check the private key checksum.
+  if (integrity_blinded_key_check(&key_derivation_key) != kHardenedBoolTrue) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // The underlying KMAC implementation is not hardened yet.
+  if (key_derivation_key.config.security_level !=
+      kOtcryptoKeySecurityLevelLow) {
+    return OTCRYPTO_NOT_IMPLEMENTED;
+  }
+
+  // Ensure `key_mode` is either KMAC-128 or KMAC-256.
+  if (key_derivation_key.config.key_mode != kOtcryptoKeyModeKdfKmac128 &&
+      key_derivation_key.config.key_mode != kOtcryptoKeyModeKdfKmac256) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Check if `key_mode` values match the input arg.
+  if (!((key_derivation_key.config.key_mode == kOtcryptoKeyModeKdfKmac128 &&
+         kmac_mode == kOtcryptoKmacModeKmac128) ||
+        (key_derivation_key.config.key_mode == kOtcryptoKeyModeKdfKmac256 &&
+         kmac_mode == kOtcryptoKmacModeKmac256))) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Validate key length of `key_derivation_key`.
+  if (key_derivation_key.config.hw_backed == kHardenedBoolTrue) {
+    return OTCRYPTO_NOT_IMPLEMENTED;
+  } else if (key_derivation_key.config.hw_backed == kHardenedBoolFalse) {
+    // Force `key_length` to be larger than minimum supported security level (
+    // 128-bit = 16 bytes).
+    if (key_derivation_key.config.key_length < 128 / 8) {
+      return OTCRYPTO_BAD_ARGS;
+    }
```

Comment:
Wouldn't this miss a too-small derivation key for KMAC256? Maybe this check could also move into the `if/else` block and be different for each `kmac_mode`.

Reply:
I think it makes sense to extend this check for each `kmac_mode` (=`security_str`). The question is, do we want to force the cryptolib user to provide a key whose length is larger than or equal to the security level? For example, if the caller uses a 128-bit key with 256-bit security, then the effective security is reduced to 128-bit. Do you think we should enforce _correct behavior_ in this sense?

Reply:
Yes, I think it makes sense to enforce the key size being at least the security level. I don't see any reason a caller would pass a smaller key except for a mistake.